sys/parsoid.js

```diff
@@ -125,109 +125,267 @@ class ParsoidService {
+     * If 'title' and 'revision' are provided, first the 'latest' bucket is checked.
+     *  Then, the stored revision is compared, if they do not equal, 404 is returned
+     *  as we can not check the stash with no tid provided.
+     *  If all the 'title', 'revision' and 'tid' are provided,
+     *  we check the latest bucket first, and then the stash bucket.
+     * @param {HyperSwitch} hyper the hyper object to rout requests
+     * @param {string} domain the domain name
+     * @param {string} title the article title
+     * @param {number} [revision] the article revision
+     * @param {string} [tid] the render TID
+     * @return {Promise<Object>} the promise that resolves to full stashed Parsoid response
+     * @private
+     */
+    _getContentWithFallback(hyper, domain, title, revision, tid) {
+        // TODO: This is temporary until we finish up the switch.
+        // The new buckets store both data-parsoid and html together,
+        // while the old buckets were storing them separately. To simplify
+        // the rest of the code and make switching the fallback easier,
+        // we make 2 requests here and emulate the new behaviour.
+        const grabContentFromOldLatestBucket = () =>
+            P.props({
+                html: hyper.get({
+                    uri: this.getOldLatestBucketURI(domain, 'html', title, revision, tid)
+                }),
+                'data-parsoid': hyper.get({
+                    uri: this.getOldLatestBucketURI(domain, 'data-parsoid', title, revision, tid)
+                })
+            })
+            .then((res) => {
+                return {
+                    status: 200,
+                    headers: res.html.headers,
+                    body: {
+                        html: res.html.body,
+                        'data-parsoid': res['data-parsoid'].body,
+                        revid: mwUtil.parseETag(res.html.headers.etag).rev
+                    }
+                };
+            });
+
+        if (!revision && !tid) {
+            return hyper.get({ uri: this.getLatestBucketURI(domain, title) })
+            .then((res) => {
+                res.body = JSON.parse(res.body.toString('utf8'));
+                return res;
+            })
+            .catch({ status: 404 }, grabContentFromOldLatestBucket);
+        } else if (!tid) {
+            return hyper.get({ uri: this.getLatestBucketURI(domain, title) })
+            .then((res) => {
+                const resEtag = mwUtil.parseETag(res.headers.etag);
+                if (revision !== resEtag.rev) {
+                    throw new HTTPError({ status: 404 });
+                }
+                res.body = JSON.parse(res.body.toString('utf8'));
+                return res;
+            })
+            .catch({ status: 404 }, grabContentFromOldLatestBucket);
+        } else {
+            return hyper.get({ uri: this.getLatestBucketURI(domain, title) })
```

Comment:
Shouldn't the logic here be: since we have both the rev and tid, check the stash bucket first, then the latest and then finally the old parsoid_bucket?